jmh-benchmarks/src/main/java/org/apache/kafka/jmh/common/FileRecordsBenchmark.java

```diff
@@ -0,0 +1,147 @@
+    private static final File fileAfter = new File("./testFileAfter.txt");
+    private static final boolean mutable = true;
+    private static final boolean fileAlreadyExists = false;
+    private static final int initFileSize = 512 * 1025 *1024;
+    private static final boolean preallocate = true;
+
+	/**
+     * Open a channel for the given file
+     * For windows NTFS and some old LINUX file system, set preallocate to true and initFileSize
+     * with one value (for example 512 * 1025 *1024 ) can improve the kafka produce performance.
+     * @param file File path
+     * @param mutable mutable
+     * @param fileAlreadyExists File already exists or not
+     * @param initFileSize The size used for pre allocate file, for example 512 * 1025 *1024
+     * @param preallocate Pre-allocate file or not, gotten from configuration.
+     */
+    private FileChannel openChannelBefore(File file,
+                                           boolean mutable,
+                                           boolean fileAlreadyExists,
+                                           int initFileSize,
+                                           boolean preallocate) throws IOException {
+        if (mutable) {
+            if (fileAlreadyExists || !preallocate) {
+                return FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.READ,
+                        StandardOpenOption.WRITE);
+            } else {
+                RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");
+                randomAccessFile.setLength(initFileSize);
+                return randomAccessFile.getChannel();
+            }
+        } else {
+            return FileChannel.open(file.toPath());
+        }
+    }
+	
+	/**
+     * Open a channel for the given file
+     * For windows NTFS and some old LINUX file system, set preallocate to true and initFileSize
+     * with one value (for example 512 * 1025 *1024 ) can improve the kafka produce performance.
+     * @param file File path
+     * @param mutable mutable
+     * @param fileAlreadyExists File already exists or not
+     * @param initFileSize The size used for pre allocate file, for example 512 * 1025 *1024
+     * @param preallocate Pre-allocate file or not, gotten from configuration.
+     */
+    private FileChannel openChannelAfter(File file,
+                                           boolean mutable,
+                                           boolean fileAlreadyExists,
+                                           int initFileSize,
+                                           boolean preallocate) throws IOException {
+        if (mutable) {
+            if (preallocate && !fileAlreadyExists) {
+                return Utils.createPreallocatedFile(file.toPath(), initFileSize);
+            } else {
+                return FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.READ,
```

Comment:
This benchmark doesn't reflect the current state right now. You can remove the benchmark changes now since we are no longer opening the file twice and hence, performance should not be a concern any more.